tests/engine/test_validation.py

```diff
@@ -0,0 +1,511 @@
+from typing import Any, Dict, Text, Type, Optional, List
+
+import pytest
+
+from rasa.engine import validation
+from rasa.engine.exceptions import GraphSchemaValidationException
+from rasa.engine.graph import GraphComponent, ExecutionContext, GraphSchema, SchemaNode
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.shared.core.domain import Domain
+from rasa.shared.nlu.training_data.training_data import TrainingData
+
+
+class TestComponentWithoutRun(GraphComponent):
+    @classmethod
+    def create(
+        cls,
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> GraphComponent:
+        return cls()
+
+
+class TestComponentWithRun(TestComponentWithoutRun):
+    def run(self) -> TrainingData:
+        pass
+
+
+def create_test_schema(
+    uses: Type,  # The unspecified type is on purpose to enable testing of invalid cases
+    constructor_name: Text = "create",
+    run_fn: Text = "run",
+    needs: Optional[Dict[Text, Text]] = None,
+    eager: bool = True,
+    parent: Optional[Type[GraphComponent]] = None,
+) -> GraphSchema:
+    parent_node = {}
+    if parent:
+        parent_node = {
+            "parent": SchemaNode(
+                needs={}, uses=parent, constructor_name="create", fn="run", config={}
+            )
+        }
+    # noinspection PyTypeChecker
+    return GraphSchema(
+        {
+            "my_node": SchemaNode(
+                needs=needs or {},
+                uses=uses,
+                eager=eager,
+                constructor_name=constructor_name,
+                fn=run_fn,
+                config={},
+            ),
+            **parent_node,
+        }
+    )
+
+
+def test_graph_component_is_no_graph_component():
+    class MyComponent:
+        def other(self) -> TrainingData:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    with pytest.raises(GraphSchemaValidationException):
```

Comment:
Is it worth checking some of the error message in all these tests just to confirm it's failing for the correct reason?

Reply:
I avoided that as you might have to adapt the tests for every little string change then 🤔 Will have a look

Reply:
done